R/drug_response_core.R

```diff
@@ -0,0 +1,463 @@
+    print(paste("Removed", removed_rows, "outliers using range_sd_filter."))
+  }
+
+  if (!is.null(remove_outliers_sd)) {
+    mean_val <- mean(all_measurements$MEASUREMENT_VALUE_HARMONIZED, na.rm = TRUE)
+    sd_val <- sd(all_measurements$MEASUREMENT_VALUE_HARMONIZED, na.rm = TRUE)
+    lower_bound <- mean_val - (remove_outliers_sd * sd_val)
+    upper_bound <- mean_val + (remove_outliers_sd * sd_val)
+
+    all_measurements <- all_measurements %>%
+      filter(.data$MEASUREMENT_VALUE_HARMONIZED >= lower_bound & .data$MEASUREMENT_VALUE_HARMONIZED <= upper_bound)
+  }
+
+  if (!is.null(winsorize_pct)) {
+    all_measurements <- all_measurements %>%
+      mutate(MEASUREMENT_VALUE_HARMONIZED = winsorize_vector(.data$MEASUREMENT_VALUE_HARMONIZED, winsorize_pct))
+  }
+
+  # 4. Filter measurements based on the time window for exposed individuals
+  time_window_years <- months_before / 12
+
+  exposed_measurements <- all_measurements %>%
+    filter(!is.na(.data$first_drug_age) & .data$time_to_drug >= 0 & .data$time_to_drug >= time_window_years)
+
+  unexposed_measurements <- all_measurements %>%
+    filter(is.na(.data$first_drug_age))
+
+  # 5. Combine and add measurement count per individual
+  final_measurements <- bind_rows(exposed_measurements, unexposed_measurements)
+
+  # Add total measurement count per individual
+  measurement_counts <- final_measurements %>%
+    group_by(.data$FINNGENID) %>%
+    summarise(n_measurements = n(), .groups = "drop")
+
+  final_measurements <- final_measurements %>%
+    left_join(measurement_counts, by = "FINNGENID")
+
+  return(final_measurements)
+}
+
+#' @title Get Median Lab Values Before Drug Purchase with MAD Outlier Removal
+#' @description Calculates the median lab value for each individual within a specified time window before
+#' the first purchase of a given drug. This function includes robust outlier removal using the
+#' Median Absolute Deviation (MAD) method and generates diagnostic plots.
+#' @param conn A `fg_data_connection` object.
+#' @param lablist A character vector of OMOP concept IDs.
+#' @param druglist A character vector of ATC drug codes.
+#' @param months_before The time window in months before the first drug purchase (default: 1).
+#' @param covariates Optional data frame for covariate data.
+#' @param covariate_cols Optional character vector of covariate column names.
+#' @param remove_outliers_mad_th The threshold for MAD-based outlier removal (default: 5).
+#' @param generate_plots Logical, whether to generate and save diagnostic plots (default: `FALSE`).
+#' @param output_dir The directory to save outputs (default: `"."`).
+#' @param output_file_prefix A prefix for output file names.
+#' @return A data frame with median lab values per individual, ready for GWAS analysis.
+#' @export
+get_median_pre_drug <- function(conn, lablist, druglist, months_before = 1,
+                                covariates = NULL, covariate_cols = NULL,
```

Comment:
same comment about the covariates.... would not add the responsibility of joining external data to these functions.